### ISSUE TITLE
A cabal.project with varying sets of tested-with.

### ISSUE THREAD

**Issue by philderbeast** [haskell-CI/haskell-ci#188]
The `uom-plugin` is a project with two packages, `uom-plugin` and `uom-plugin-examples`. For the `.travis.yml` setup, I'd like to build the examples for only a subset of those ghc versions that I build the plugin with, as in commit https://github.com/BlockScope/uom-plugin/commit/5266aee1aab831f33b073e777ebbad6069b4b705 for issue https://github.com/adamgundry/uom-plugin/issues/59.

This is what 'd like;

```
-- uom-plugin
tested-with: GHC == 8.0.2, GHC == 8.2.2, GHC == 8.4.3 GHC == 8.6.1
-- uom-plugin-examples
tested-with: GHC == 8.0.2, GHC == 8.2.2

[nix-shell]$ runhaskell ../haskell-ci/make_travis_yml_2.hs ./cabal.project > .travis.yml
*ERROR* uom-plugin-examples is missing tested-with annotations for: 8.4.3,8.6.1
```

When the `tested-with` attribute is empty or missing from the examples this is the error;

```
-- uom-plugin
tested-with: GHC == 8.0.2, GHC == 8.2.2, GHC == 8.4.3 GHC == 8.6.1
-- uom-plugin-examples
tested-with:

[nix-shell]$ runhaskell ../haskell-ci/make_travis_yml_2.hs ./cabal.project > .travis.yml
*ERROR* empty or missing top-level 'tested-with:'
definition in uom-plugin-examples/uom-plugin-examples.cabal file; example definition:

tested-with: GHC==8.6.1, GHC==8.4.4, GHC==8.2.2, GHC==8.0.2, GHC==7.10.3,
GHC==7.8.4, GHC==7.6.3, GHC==7.4.2, GHC==7.2.2, GHC==7.0.4
```

If I set the examples to have the very same set of `tested-with` versions as the plugin then a `.travis.yml` is generated;

```
-- uom-plugin
tested-with: GHC == 8.0.2, GHC == 8.2.2, GHC == 8.4.3 GHC == 8.6.1
-- uom-plugin-examples
tested-with: GHC == 8.0.2, GHC == 8.2.2, GHC == 8.4.3 GHC == 8.6.1

[nix-shell]$ runhaskell ../haskell-ci/make_travis_yml_2.hs ./cabal.project > .travis.yml
*INFO* Generating Travis-CI config for testing for GHC versions: 8.0.2 8.2.2 8.4.3 8.6.1
```

In case of projects having matching sets of `tested-with` versions, the generated `.travis.yml` deletes and then regenerates the `cabal.project` and builds from this;

```
rm -fv cabal.project cabal.project.local
...
"printf 'packages: \"uom-plugin\" \"uom-plugin-examples\"\\n' > cabal.project"
...
cabal new-build -w ${HC} ${TEST} ${BENCH} --project-file="./cabal.project" --dep -j2 all
```

When the packages in a project don't have matching sets of `tested-with` versions would it be possible to generate a `.travis.yml` that builds all of those sets of packages?

**Comment by phadej** [comment]
Out of curiosity, why you don't build examples with all versions?

But seriously, yes, it would be nice to 
- have some packages with different `tested-with` (e.g. extra packages supporting only newer GHC)
- have some packages without `tested-with` (examples not meant to be released, but tested with everything)

The problem is how to make these explicit choices, i.e. command line UI, ideas? I want to preserve the default that all `tested-with`, so I don't accidentally forgot to update it in some packages (case: `servant`)

**Comment by philderbeast** [comment]
The plugin works as intended when `GHC == 8.2.2`. I've got the plugin itself building with more recent versions of ghc but for code using this plugin, doing `-fplugin Data.UnitsOfMeasure.Plugin`, we see typechecking errors with;

> lots of unsolved constraints involving flattening skolems
https://github.com/adamgundry/uom-plugin/issues/43

This is the problem with the `uom-plugin-examples` package right now and why I'd like to exclude `GHC == 8.4.3` and later from its `tested-with` set.

**Comment by philderbeast** [comment]
I can mask off the tests and the examples by adding conditionals to the `.cabal` files;

```
  if impl(ghc > 8.2.2)
    buildable: False
```

I did this upstream of the `.cabal` files with hpack's `package.yaml` files as in https://github.com/sol/hpack/issues/285.

**Comment by chris-martin** [comment]
Same problem here - I'm trying to set up CI for [partial-semigroup](https://github.com/chris-martin/partial-semigroup). That repo contains `partial-semigroup` which supports a ton of old GHC versions, and `partial-semigroup-hedgehog` which has a `hedgehog` dependency and won't build earlier than `7.10.3`. I'll probably end up dropping support for older GHCs in `partial-semigroup` just because I can't figure out how to get the CI working.

**Comment by phadej** [comment]
If one specifieds `--job-selection=any` (vs. default `uniform`) then whatever `tested-with` combinations there is, generated script will try to replicate it.

**Comment by phadej** [comment]
Solved by #217
